torch/csrc/Storage.cpp

```diff
@@ -84,15 +84,124 @@ PyObject* THPStorage_NewWithStorage(
   return obj;
 }
 
+// This object is used as the `ctx` argument for DataPtr to implement a shared
+// DataPtr. Normally, a DataPtr is unique, but we use this custom context and
+// the deleter function below to make the DataPtr act like a non-unique
+// DataPtr. This context object holds onto an inner context and deleter
+// function which handle the actual deletion of the data when the refcount
+// reaches 0.
+//
+// This shared DataPtr feature is only used when storages are shared between
+// multiple Python interpreters in torch::deploy (MultiPy). Before storages had
+// PyObject preservation, interpreters could just share the same StorageImpl
+// instance. But now a StorageImpl can only be associated with one interpreter
+// in order to properly manage a zombie PyObject. So we share storages across
+// Python interpreters by creating a different StorageImpl instance for each
+// one, but they all point to the same data.
+struct RefcountedDeleterContext {
+  RefcountedDeleterContext(void* other_ctx, c10::DeleterFnPtr other_deleter)
+      : other_ctx(other_ctx, other_deleter), refcount(1) {}
+
+  void incref() {
+    refcount++;
+  }
+
+  std::unique_ptr<void, c10::DeleterFnPtr> other_ctx;
+  size_t refcount;
+};
+
+// This function is used as the `ctx_deleter` for DataPtr to implement a
+// shared DataPtr.
+// Warning: This should only be called on a RefcountedDeleterContext that was
+// allocated on the heap with `new`
+void refcounted_deleter(void* ctx_) {
+  RefcountedDeleterContext& ctx =
+      *reinterpret_cast<RefcountedDeleterContext*>(ctx_);
+  ctx.refcount--;
+  if (ctx.refcount == 0) {
+    ctx.other_ctx = nullptr;
+    delete &ctx;
+  }
+}
+
+// If the storage has a normal unique DataPtr, replace it with a DataPtr
+// that can be shared between multiple StorageImpls
+void maybeReplaceDataPtrWithShared(c10::Storage storage) {
+  c10::DataPtr& data_ptr = storage.mutable_data_ptr();
+
+  if (data_ptr.get_deleter() == &refcounted_deleter) {
```

Comment:
This does not work as I expected. To determine if we need to replace the data pointer, we first need to find out if it's is already shared. I thought I could just check if the deleter function is `refcounted_deleter`. However, it turns out that different threads in MultiPy map the PyTorch library to different memory addresses, and the value of `&refcounted_deleter` is different for each thread. I'm trying to think if there's currently any way at all to find out if the data pointer is shared, and I think the answer is no. The only other thing that a shared `DataPtr` holds is a `void*` to a `RefcountedDeleterContext` object, but there's no way to tell what type of object a `void*` points to.

So I'm pretty sure I need to figure out how to replace the storage's data pointer with a shared one in the main thread before the storage is given out to other threads. Then any time `has_pyobj && !is_same_interpreter` (in `THPStorage_Wrap`), we'll just assume that it's already a shared data pointer. I'm not sure how to do this yet, I think it might require a change to MultiPy

Another option is to add a flag on `DataPtr` to say whether it's deleter function is `refcounted_deleter` or not. But at that rate, I guess it could potentially make more sense to just explicitly support a shared pointer within `DataPtr`

Reply:
Why can't you move refcounted_deleter to libtorch so it is shared everywhere?

Reply:
The entire library is mapped to a different address space for each thread. For instance, I added this print to find the address of `torch::empty_like` in two different threads created by multipy:

```
std::cout << "  &torch::empty_like: " << (void*)&torch::empty_like << std::endl;
```

One thread printed this:
```
  &torch::empty_like: 0x7fd58800c6b0
```

The other thread printed this:
```
  &torch::empty_like: 0x7fd5902cd6b0
```

Reply:
> The entire library is mapped to a different address space for each thread

Only libtorch_python. At a guess, empty_like is in libtorch_python for reasons. Try something like at::empty, that should be stable.

Reply:
Ah ok, evidently `at::empty`'s address isn't stable, but things in `c10`, like `c10::GetAllocator`, appear to be.
```cpp
    size_t thread_id = std::hash<std::thread::id>{}(std::this_thread::get_id());
    at::Tensor (*func0)(at::IntArrayRef, at::TensorOptions, c10::optional<at::MemoryFormat> memory_format) = torch::empty;
    at::Tensor (*func1)(at::IntArrayRef, at::TensorOptions, c10::optional<at::MemoryFormat> memory_format) = at::empty;
    void* func2 = (void*)&c10::GetAllocator;
    std::cout << "  thread: " << thread_id << ",  func0: " << (void*)func0 << std::endl;
    std::cout << "  thread: " << thread_id << ",  func1: " << (void*)func1 << std::endl;
    std::cout << "  thread: " << thread_id << ",  func2: " << (void*)func2 << std::endl;
```

Output
```
  thread: 17112257896192799938,  func0: 0x7f6d411f8c90
  thread: 17112257896192799938,  func1: 0x7f6d411f7d80
  thread: 17112257896192799938,  func2: 0x7f6dd8254720
  thread: 3004227773937621798,  func0: 0x7f6d38f37c90
  thread: 3004227773937621798,  func1: 0x7f6d38f36d80
  thread: 3004227773937621798,  func2: 0x7f6dd8254720
```

So I'll move the refcounted deleter stuff to `c10` (which is arguably a better place anyway) and it should work

Reply:
That worked